compiler/src/main/java/bullet/impl/ComponentProcessingStep.java

```diff
@@ -114,35 +124,32 @@ private void generateObjectGraph(TypeElement element) {
           break;
         case SIMPLE_MEMBERS_INJECTION:
         case MEMBERS_INJECTOR:
-          membersInjectionMethods.add(methodDescriptor);
+          memberInjectorsTypeGraph.addVertex(methodDescriptor);
+          memberInjectionMethodsTypeDescriptorMap.put(methodDescriptor.type(), methodDescriptor);
+
+          for (ComponentMethodDescriptor pendingChild : potentialPendingEdges.get(methodDescriptor.type())) {
+            memberInjectorsTypeGraph.addEdge(pendingChild, methodDescriptor);
+          }
+
+          DeclaredType declaredType = (DeclaredType) methodDescriptor.type();
+          TypeElement typeElement = (TypeElement) declaredType.asElement();
+
+          List<TypeMirror> parents = new ArrayList<>(typeElement.getInterfaces());
+          parents.add(typeElement.getSuperclass());
+          for (TypeMirror parent : parents) {
+            ComponentMethodDescriptor parentDescriptor = memberInjectionMethodsTypeDescriptorMap.get(parent);
+            if (parentDescriptor == null) {
+              potentialPendingEdges.put(parent, methodDescriptor);
```

Comment:
How about walking the graph and, for each existing vertex:

 * if it's the same type, skip (no need to track duplicates, only one will be called in the end anyway)
 * if it's a subtype, look at existing outbound edges:
    * if the other vertex is the same type, skip
    * if it's a super type, then insert the type between the 2 vertices
    * if no vertex matches the above rules, insert a new outbound edge to the type
 * similarly for supertypes and inbound edges

Or something like that.